package NYT.NHydra.NProto;

import "yt/core/misc/proto/guid.proto";
import "yt/core/rpc/proto/rpc.proto";

////////////////////////////////////////////////////////////////////////////////

message TPeerKindExt
{
    extend NRpc.NProto.TReqDiscover
    {
        optional TPeerKindExt peer_kind_ext = 100;
    }

    required int32 peer_kind = 1;
}

////////////////////////////////////////////////////////////////////////////////

message TReqReadChangeLog
{
    required int32 changelog_id = 1;
    required int32 start_record_id = 2;
    required int32 record_count = 3;
}

message TRspReadChangeLog
{
    // Changelog records are passed in the attachment and are packed with PackRefs.
}

////////////////////////////////////////////////////////////////////////////////

message TReqLookupChangelog
{
    required int32 changelog_id = 1;
}

message TRspLookupChangelog
{
    optional int32 record_count = 2;
}
                                                                              
////////////////////////////////////////////////////////////////////////////////

message TReqPingFollower
{
    required NYT.NProto.TGuid epoch_id = 1;
    required int64 committed_revision = 2;
    required int64 ping_revision = 3;
}

message TRspPingFollower
{
    required int32 state = 1;
}
                                                                              
////////////////////////////////////////////////////////////////////////////////

message TReqAcceptMutations
{
    required NYT.NProto.TGuid epoch_id = 1;
    required int64 start_revision = 2;
    required int64 committed_revision = 3;
    // Records data is passed via attachments.
}

message TRspAcceptMutations
{
    required bool logged = 1;
}
                                                                              
////////////////////////////////////////////////////////////////////////////////

message TReqBuildSnapshot
{
    required NYT.NProto.TGuid epoch_id = 1;
    required int64 revision = 2;
    optional bool set_read_only = 3 [default = false];
}

message TRspBuildSnapshot
{
    required uint64 checksum = 1;
}
                                                                              
////////////////////////////////////////////////////////////////////////////////

message TReqRotateChangelog
{
    required NYT.NProto.TGuid epoch_id = 1;
    required int64 revision = 2;
}

message TRspRotateChangelog
{ }
                                                                              
////////////////////////////////////////////////////////////////////////////////

message TReqSyncWithLeader
{
    required NYT.NProto.TGuid epoch_id = 1;
}

message TRspSyncWithLeader
{
    required int64 committed_revision = 1;
}

////////////////////////////////////////////////////////////////////////////////

message TReqCommitMutation
{
    required string type = 1;
    optional NYT.NProto.TGuid mutation_id = 2;
    optional bool retry = 3 [default = false];
    // Request data is passed via attachment.
}

message TRspCommitMutation
{
    // Response data is passed via attachments.
}

////////////////////////////////////////////////////////////////////////////////

message TReqForceBuildSnapshot
{
    optional bool set_read_only = 1 [default = false];
}

message TRspForceBuildSnapshot
{
    required int32 snapshot_id = 1;
}

////////////////////////////////////////////////////////////////////////////////
